File: neuroops-agent/sdk/python/neuroopssdk/clients/snmpclient.py
#  NeurOps Agent — Python SDK SNMP Client
#  Copyright (c) NeurOps 2025. All rights reserved.
#
#  Wraps pysnmp for SNMP v1/v2c/v3 GET, GETNEXT, WALK, BULK, and SET.
#  Used by network device metric plugins (routers, switches, firewalls,
#  UPS, storage arrays, printers, and any SNMP-managed device).
#
#  Usage:
#    client = SNMPClient(context, logger)
#    result = client.get(["1.3.6.1.2.1.1.1.0", "1.3.6.1.2.1.1.5.0"])
#    result = client.walk("1.3.6.1.2.1.2.2")     # IF-MIB interfaces

from pysnmp.hlapi import (
    CommunityData,
    ContextData,
    Integer32,
    ObjectIdentity,
    ObjectType,
    OctetString,
    SnmpEngine,
    UdpTransportTarget,
    UsmUserData,
    authHMACMD5Protocol,
    authHMACSHAProtocol,
    getCmd,
    nextCmd,
    setCmd,
    bulkCmd,
    usmDESPrivProtocol,
    usmAesCfb128Protocol,
    usmNoAuthProtocol,
    usmNoPrivProtocol,
)

from neuroopssdk.constants import Constant


class SNMPClient:

    # SNMP versions
    V1  = "v1"
    V2C = "v2c"
    V3  = "v3"

    # SNMPv3 auth/priv protocols
    AUTH_MD5    = "MD5"
    AUTH_SHA    = "SHA"
    PRIV_DES    = "DES"
    PRIV_AES128 = "AES128"

    DEFAULT_PORT    = 161
    DEFAULT_TIMEOUT = 5
    DEFAULT_RETRIES = 1

    def __init__(self, context: dict, logger):
        self._log      = logger
        self._host     = context.get(Constant.OBJECT_IP, "")
        self._port     = int(context.get("snmp.port", self.DEFAULT_PORT))
        self._version  = context.get("snmp.version", self.V2C).lower()
        self._timeout  = int(context.get("snmp.timeout", self.DEFAULT_TIMEOUT))
        self._retries  = int(context.get("snmp.retries", self.DEFAULT_RETRIES))

        # v1/v2c
        self._community = context.get("snmp.community", "public")

        # v3
        self._username     = context.get(Constant.USERNAME, "")
        self._auth_proto   = context.get("snmp.auth.protocol", self.AUTH_SHA)
        self._auth_pass    = context.get("snmp.auth.password", "")
        self._priv_proto   = context.get("snmp.priv.protocol", self.PRIV_AES128)
        self._priv_pass    = context.get("snmp.priv.password", "")
        self._context_name = context.get("snmp.context", "")

        self._engine    = SnmpEngine()

    # ── Public API ─────────────────────────────────────────────────────────

    def get(self, oids: list) -> dict:
        """
        Perform SNMP GET for a list of OIDs.
        Returns: {"status": "succeed", "result": {"1.3.6.1...": "value", ...}}
        """
        try:
            result = {}
            objects = [ObjectType(ObjectIdentity(oid)) for oid in oids]
            error_indication, error_status, error_index, var_binds = next(
                getCmd(
                    self._engine,
                    self._auth_data(),
                    self._transport(),
                    ContextData(),
                    *objects,
                )
            )
            if error_indication:
                return self._fail(str(error_indication), Constant.ERROR_NO_RESPONSE)
            if error_status:
                return self._fail(
                    f"SNMP error at {error_index}: {error_status.prettyPrint()}",
                    Constant.ERROR_BAD_RESPONSE,
                )
            for var_bind in var_binds:
                oid, value = var_bind
                result[str(oid)] = self._decode_value(value)
            return {"status": Constant.STATUS_SUCCEED, "result": result}
        except Exception as e:
            return self._fail(str(e), Constant.ERROR_INTERNAL)

    def walk(self, base_oid: str) -> dict:
        """
        Perform SNMP WALK starting at base_oid.
        Returns: {"status": "succeed", "result": {<oid>: <value>, ...}}
        """
        try:
            result = {}
            for error_indication, error_status, _, var_binds in nextCmd(
                self._engine,
                self._auth_data(),
                self._transport(),
                ContextData(),
                ObjectType(ObjectIdentity(base_oid)),
                lexicographicMode=False,
            ):
                if error_indication:
                    return self._fail(str(error_indication), Constant.ERROR_NO_RESPONSE)
                if error_status:
                    return self._fail(str(error_status.prettyPrint()), Constant.ERROR_BAD_RESPONSE)
                for var_bind in var_binds:
                    oid, value = var_bind
                    result[str(oid)] = self._decode_value(value)
            return {"status": Constant.STATUS_SUCCEED, "result": result}
        except Exception as e:
            return self._fail(str(e), Constant.ERROR_INTERNAL)

    def bulk_walk(self, base_oid: str, max_repetitions: int = 25) -> dict:
        """
        Perform SNMP BULK WALK (v2c/v3 only).
        More efficient than walk() for large tables.
        """
        try:
            result = {}
            for error_indication, error_status, _, var_binds in bulkCmd(
                self._engine,
                self._auth_data(),
                self._transport(),
                ContextData(),
                0, max_repetitions,
                ObjectType(ObjectIdentity(base_oid)),
                lexicographicMode=False,
            ):
                if error_indication:
                    return self._fail(str(error_indication), Constant.ERROR_NO_RESPONSE)
                if error_status:
                    break
                for var_bind in var_binds:
                    oid, value = var_bind
                    result[str(oid)] = self._decode_value(value)
            return {"status": Constant.STATUS_SUCCEED, "result": result}
        except Exception as e:
            return self._fail(str(e), Constant.ERROR_INTERNAL)

    def set(self, oid: str, value, value_type: str = "string") -> dict:
        """
        Perform SNMP SET.
        value_type: "string" | "integer" | "oid"
        """
        try:
            if value_type == "integer":
                snmp_value = Integer32(int(value))
            else:
                snmp_value = OctetString(str(value))

            error_indication, error_status, error_index, _ = next(
                setCmd(
                    self._engine,
                    self._auth_data(),
                    self._transport(),
                    ContextData(),
                    ObjectType(ObjectIdentity(oid), snmp_value),
                )
            )
            if error_indication:
                return self._fail(str(error_indication), Constant.ERROR_NO_RESPONSE)
            if error_status:
                return self._fail(str(error_status.prettyPrint()), Constant.ERROR_BAD_RESPONSE)
            return {"status": Constant.STATUS_SUCCEED}
        except Exception as e:
            return self._fail(str(e), Constant.ERROR_INTERNAL)

    def get_table(self, column_oids: list) -> dict:
        """
        Walk multiple OID columns in parallel and assemble as a table.
        Returns: {"status": "succeed", "result": {<index>: {<oid>: <val>}}}
        """
        table = {}
        for oid in column_oids:
            r = self.walk(oid)
            if r["status"] != Constant.STATUS_SUCCEED:
                return r
            for full_oid, value in r["result"].items():
                # Extract table index (last component)
                idx = full_oid.rsplit(".", 1)[-1]
                table.setdefault(idx, {})[oid] = value
        return {"status": Constant.STATUS_SUCCEED, "result": table}

    # ── Helpers ────────────────────────────────────────────────────────────

    def _auth_data(self):
        if self._version in (self.V1, self.V2C):
            mp_model = 0 if self._version == self.V1 else 1
            return CommunityData(self._community, mpModel=mp_model)

        # SNMPv3
        auth_protocol = {
            self.AUTH_MD5: authHMACMD5Protocol,
            self.AUTH_SHA: authHMACSHAProtocol,
        }.get(self._auth_proto.upper(), usmNoAuthProtocol)

        priv_protocol = {
            self.PRIV_DES:    usmDESPrivProtocol,
            self.PRIV_AES128: usmAesCfb128Protocol,
        }.get(self._priv_proto.upper(), usmNoPrivProtocol)

        return UsmUserData(
            self._username,
            authKey=self._auth_pass or None,
            privKey=self._priv_pass or None,
            authProtocol=auth_protocol,
            privProtocol=priv_protocol,
        )

    def _transport(self):
        return UdpTransportTarget(
            (self._host, self._port),
            timeout=self._timeout,
            retries=self._retries,
        )

    @staticmethod
    def _decode_value(value) -> any:
        """Convert pysnmp value objects to native Python types."""
        try:
            # Try integer first
            return int(value)
        except Exception:
            pass
        try:
            # Try string
            s = value.prettyPrint()
            # Strip hex prefix if pysnmp added it
            if s.startswith("0x"):
                try:
                    return bytes.fromhex(s[2:]).decode(errors="replace")
                except Exception:
                    pass
            return s
        except Exception:
            return str(value)

    def _fail(self, message: str, code: str = Constant.ERROR_INTERNAL) -> dict:
        self._log.errorf("SNMPClient [%s]: %s", self._host, message)
        return {
            "status":     Constant.STATUS_FAIL,
            "error":      message,
            "error.code": code,
        }

File: neuroops-agent/sdk/go/clients/snmp/snmpclient.go
/*
 * NeurOps Agent — Go SDK SNMP Client
 * Copyright (c) NeurOps 2025. All rights reserved.
 *
 * Wraps gosnmp for SNMP v1/v2c/v3 GET, WALK, BULK, and SET.
 * Used by network device metric plugins.
 *
 * Usage:
 *   client := snmp.New(context, logger)
 *   if err := client.Connect(); err != nil { ... }
 *   defer client.Disconnect()
 *   result, err := client.Get([]string{"1.3.6.1.2.1.1.1.0"})
 */

package snmp

import (
	"fmt"
	"strings"
	"time"

	g "github.com/gosnmp/gosnmp"

	"github.com/neuroops/agent/internal/logger"
	"github.com/neuroops/agent/sdk/go/types"
)

// Client wraps a gosnmp session.
type Client struct {
	session *g.GoSNMP
	log     *logger.Logger
}

// New creates an SNMP Client from a plugin context map.
func New(ctx types.NeurOpsMap, log *logger.Logger) *Client {
	port := uint16(ctx.GetInt("snmp.port"))
	if port == 0 {
		port = 161
	}

	timeout := time.Duration(ctx.GetInt("snmp.timeout")) * time.Second
	if timeout == 0 {
		timeout = 5 * time.Second
	}

	retries := int(ctx.GetInt("snmp.retries"))
	if retries == 0 {
		retries = 1
	}

	version := resolveVersion(ctx.GetString("snmp.version"))
	community := ctx.GetString("snmp.community")
	if community == "" {
		community = "public"
	}

	session := &g.GoSNMP{
		Target:    ctx.GetString("object.ip"),
		Port:      port,
		Version:   version,
		Community: community,
		Timeout:   timeout,
		Retries:   retries,
		MaxOids:   g.MaxOids,
	}

	// SNMPv3 parameters
	if version == g.Version3 {
		session.SecurityModel = g.UserSecurityModel
		session.MsgFlags      = resolveSecLevel(ctx)
		session.SecurityParameters = &g.UsmSecurityParameters{
			UserName:                 ctx.GetString("username"),
			AuthenticationProtocol:  resolveAuthProto(ctx.GetString("snmp.auth.protocol")),
			AuthenticationPassphrase: ctx.GetString("snmp.auth.password"),
			PrivacyProtocol:          resolvePrivProto(ctx.GetString("snmp.priv.protocol")),
			PrivacyPassphrase:        ctx.GetString("snmp.priv.password"),
		}
	}

	return &Client{session: session, log: log}
}

// Connect opens the SNMP session.
func (c *Client) Connect() error {
	return c.session.Connect()
}

// Disconnect closes the session.
func (c *Client) Disconnect() {
	if c.session.Conn != nil {
		_ = c.session.Conn.Close()
	}
}

// Get performs SNMP GET for the given OIDs.
// Returns a map of OID → value.
func (c *Client) Get(oids []string) (map[string]any, error) {
	result, err := c.session.Get(oids)
	if err != nil {
		return nil, fmt.Errorf("SNMP GET: %w", err)
	}
	return c.decodeVarBinds(result.Variables), nil
}

// Walk performs SNMP WALK starting at baseOID.
func (c *Client) Walk(baseOID string) (map[string]any, error) {
	result := make(map[string]any)
	err := c.session.Walk(baseOID, func(pdu g.SnmpPDU) error {
		result[pdu.Name] = decodeValue(pdu)
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("SNMP WALK %s: %w", baseOID, err)
	}
	return result, nil
}

// BulkWalk performs SNMP BULK WALK (v2c/v3 only).
func (c *Client) BulkWalk(baseOID string) (map[string]any, error) {
	result := make(map[string]any)
	err := c.session.BulkWalk(baseOID, func(pdu g.SnmpPDU) error {
		result[pdu.Name] = decodeValue(pdu)
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("SNMP BULKWALK %s: %w", baseOID, err)
	}
	return result, nil
}

// GetTable walks multiple OID columns and assembles them into a table
// keyed by row index (last component of each OID).
func (c *Client) GetTable(columnOIDs []string) (map[string]map[string]any, error) {
	table := make(map[string]map[string]any)
	for _, oid := range columnOIDs {
		rows, err := c.Walk(oid)
		if err != nil {
			return nil, err
		}
		for fullOID, value := range rows {
			idx := lastComponent(fullOID)
			if table[idx] == nil {
				table[idx] = make(map[string]any)
			}
			table[idx][oid] = value
		}
	}
	return table, nil
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func (c *Client) decodeVarBinds(pdus []g.SnmpPDU) map[string]any {
	result := make(map[string]any, len(pdus))
	for _, pdu := range pdus {
		result[pdu.Name] = decodeValue(pdu)
	}
	return result
}

func decodeValue(pdu g.SnmpPDU) any {
	switch pdu.Type {
	case g.Integer, g.Counter32, g.Gauge32, g.TimeTicks,
		g.Counter64, g.Uinteger32:
		switch v := pdu.Value.(type) {
		case uint:
			return int64(v)
		case uint32:
			return int64(v)
		case uint64:
			return int64(v)
		case int:
			return int64(v)
		default:
			return pdu.Value
		}
	case g.OctetString:
		if b, ok := pdu.Value.([]byte); ok {
			return string(b)
		}
		return fmt.Sprintf("%v", pdu.Value)
	case g.ObjectIdentifier:
		return fmt.Sprintf("%v", pdu.Value)
	default:
		return pdu.Value
	}
}

func lastComponent(oid string) string {
	parts := strings.Split(strings.TrimPrefix(oid, "."), ".")
	if len(parts) == 0 {
		return oid
	}
	return parts[len(parts)-1]
}

func resolveVersion(v string) g.SnmpVersion {
	switch strings.ToLower(strings.TrimPrefix(v, "v")) {
	case "1":
		return g.Version1
	case "3":
		return g.Version3
	default:
		return g.Version2c
	}
}

func resolveSecLevel(ctx types.NeurOpsMap) g.SnmpV3MsgFlags {
	authPass := ctx.GetString("snmp.auth.password")
	privPass := ctx.GetString("snmp.priv.password")
	if authPass != "" && privPass != "" {
		return g.AuthPriv
	}
	if authPass != "" {
		return g.AuthNoPriv
	}
	return g.NoAuthNoPriv
}

func resolveAuthProto(proto string) g.SnmpV3AuthProtocol {
	switch strings.ToUpper(proto) {
	case "MD5":
		return g.MD5
	default:
		return g.SHA
	}
}

func resolvePrivProto(proto string) g.SnmpV3PrivProtocol {
	switch strings.ToUpper(proto) {
	case "DES":
		return g.DES
	default:
		return g.AES
	}
}
